Introduce a SPACK_PYTHON environment variable

chuckatkins opened the pull request:
The `SPACK_PYTHON` environment variable can be set to a python interpreter to be used by the `spack` command.  This allows the `spack` command itself to use a consistent and separate interpreter from whatever python installation might be used for package building.

One example of the use case for this would be in RHEL 8 / CentOS 8 to use the system python for the spack command:
```
$ export SPACK_PYTHON=/usr/libexec/platform-python
$ . share/spack/setup-env.sh
$ spack load python@3.8.6
$ python --version
Python 3.8.6
$ spack python --version
Python 3.6.8
$ spack unload python@3.8.6
$ spack load python@3.7.6
$ python --version
Python 3.7.6
$ spack python --version
Python 3.6.8
$
```

Here we have the `spack` command always use the stable system provided python 3.6.8 in `/usr/libexec/platform-python`, the same use used by the system's `dnf` command, while continuing to be able to load and unload different versions of python from the current environment for use with packages.

tgamblin commented:
I like this!  Two things:

- [ ] Can you add docs?
- [ ] Should we add the RHEL8 platform python to the list of pythons to use?  Maybe at the end of the list?

I haven't tried adding that python to spack because I'm a little scared of the "no guarantees" thing, and IIRC the tests didn't pass with that python.  Have you tried them?

adamjstewart commented:
I love it! I might even use this on all of my systems. Currently I use Spack to install my environments, which contain Python. This causes all kinds of problems when I need to update my environment, as the Python I'm running Spack with disappears mid-installation. Another thing worth considering is if there's a way to disable `PYTHONPATH` when running Spack. I often find that if I `spack load py-foo` and the Python version is different from the one I'm running Spack with, all kinds of things break.

What is `platform-python`? We should add that prefix to the automatic detection in the `python` package.

tgamblin commented:
`platform-python` is the `python` that runs DNF in RHEL8, but that's all they guarantee it for.  So they may strip it down, remove parts, etc. and it may break.  I like the idea of being able to use it; I just think we would need to keep more of an eye on it since it guarantees only that DNF will run

chuckatkins commented:
> * [ ]  Can you add docs?

Certainly!  I wasn't sure where to add it though.

> * [ ]  Should we add the RHEL8 platform python to the list of pythons to use?  Maybe at the end of the list?

Perhaps, I'd like to get some more use on it first though so we don't pickup a python that only "kind of works".  With the way it's set here it requires an explicit opt in.

> I haven't tried adding that python to spack because I'm a little scared of the "no guarantees" thing, and IIRC the tests didn't pass with that python. Have you tried them?

I haven't.  I'll run through the tests though and see.  If they do pass then it should be safe to put `/usr/libexec/platform-python` in the search path.  In particular I started using this to setup spack in the RHEL8 Universal Base Image with almost no additional system packages aside from git, curl, gcc, gcc-c++, gcc-gfortran, gzip, bzip2, xz, and unzip.

chuckatkins commented:
Okay, looking through docs and RedHat blogs, I'm confident that we should *not* include `platform-python` in the search:

>  It only includes the parts of Python needed for the system to function, and there are no guarantees that any particular feature won’t be removed from it in the future.

However, it may be useful to perform the search in the `setup-env.sh` as well to set `SPACK_PYTHON` if not set.  This would give the additional benefit of a fixed interpreter used by spack without requiring it be explicitly set.

tgamblin commented:
> Okay, looking through docs and RedHat blogs, I'm confident that we should not include platform-python in the search:

Per our call just now -- let's add it and add a RHEL UBI container to CI to test with `platform-python`.  If it fails one day, we can remove it, but I think if we monitor it via CI we can ensure it keeps working.  If it is last in the list of preferred pythons, it should be easy for a user to either install a python (which would take precedence) or set `SPACK_PYTHON`

adamjstewart commented:
Are there systems that have `platform-python` but not `python2` or `python3`?

chuckatkins commented:
> Are there systems that have `platform-python` but not `python2` or `python3`?

A minimal EL8 install

Commit:
Introduce a SPACK_PYTHON environment variable

The SPACK_PYTHON environment variable can be set to a python interpreter to be
used by the spack command.  This allows the spack command itself to use a
consistent and separate interpreter from whatever python might be used for package
building.

chuckatkins commented:
> > Okay, looking through docs and RedHat blogs, I'm confident that we should not include platform-python in the search:
> 
> Per our call just now -- let's add it and add a RHEL UBI container to CI to test with `platform-python`. If it fails one day, we can remove it, but I think if we monitor it via CI we can ensure it keeps working. If it is last in the list of preferred pythons, it should be easy for a user to either install a python (which would take precedence) or set `SPACK_PYTHON`

If it's okay by you, I'd suggest adding `platform-python` and the ubi image in a separate PR as the `SPACK_PYTHON` feature can really stand alone with out it.

This feature should be good to go now.

tgamblin commented on share/spack/setup-env.sh:
```diff
@@ -326,6 +326,16 @@ if [ "$_sp_shell" = bash ]; then
+if [ -z "${SPACK_PYTHON}" ]; then
+    for cmd in python3 python python2; do
+        if command -v > /dev/null $cmd; then
+            export SPACK_PYTHON="$(which $cmd)"
```
I think this use of `which` can break in `zsh`, where default `which` will print an alias (not a path) if one is set.  I am not sure how likely a complex python alias is, but I've seen some things in user configs ,so I think `command -v` is better here.

I am not sure whether this is also an issue in fish.  Maybe @JBlaschke knows.

JBlaschke replied:
Fish is unusual in that `alias` defines a `function` type (you can try this yourself my creating an alias, and then using `type` and `functions` to find out what it is -- the outputs of both is the same => the alias is actually a function).

Since `which` doesn't know about functions, in the case of fish it won't print out aliases. The current solution with`command -s ` is an elegant one though, as `which` is not a fish built-in.

chuckatkins replied:
For sh the bash `command` builtin is used; csh doesn't have a builtin so it uses the `/usr/bin/command` system binary, both of which use the same `-v` argument.  The fish changes use fish's `command` builtin, which uses a different `-s` argument than the bash builtin.

JBlaschke replied:
Right! Sorry, the `-s` is not strictly necessary in:
> For basic compatibility with POSIX command, the -v flag is recognized as an alias for -s.
(from fish's `command` manpages)

So you can use `command -v` on fish also. I tested this on version 3.1.2

tgamblin commented on share/spack/setup-env.sh:
```diff
@@ -326,6 +326,16 @@ if [ "$_sp_shell" = bash ]; then
 fi
 
+# Identify and lock the python interpreter
+if [ -z "${SPACK_PYTHON}" ]; then
```
This will fail when SPACK_PYTHON is not set if users use `set -e` in their shells (or in scripts where it's set).  Can you use the "idiom" (aka horribly unintuitive sh-ism) we use above, i.e.:

```
if [ -n "${_sp_initializing:-}" ]; then
...
```

adamjstewart commented on bin/spack:
```diff
@@ -10,9 +10,9 @@
-for cmd in python3 python python2; do
-   command -v > /dev/null $cmd && exec $cmd $0 "$@"
+# prefer SPACK_PYTHON environment variable, python3, python, then python2
+for cmd in "$SPACK_PYTHON" python3 python python2; do
```
```suggestion
for cmd in $SPACK_PYTHON python3 python python2; do
```
The difference is that if `SPACK_PYTHON` isn't set, we loop over 3 commands instead of 4 ("" python3 python python2)

chuckatkins replied:
I considered that but wanted to be able to allow spaces in the path so the quotes are needed.  In the case of an empty string `command -v > /dev/null ""` will fail and iteration will proceed to `python3`.  However, with the addition of the `setup-env.*sh` changes the use case for an empty string will me much less common.

chuckatkins commented:
Well that didn't go as expected...  give me a minute and I'll fix this.

chuckatkins commented:
Finally, all the shell tests pass.

adamjstewart commented:
What is the motivation for having the setup scripts set `SPACK_PYTHON`?

chuckatkins commented:
> What is the motivation for having the setup scripts set `SPACK_PYTHON

After initializing the environment with an appropriate shell setup script then the spack command will use a consistent python interpreter for the duration of the environment, even if you load a different python with `spack load` for use with applications.

chuckatkins commented:
Ping @tgamblin

chuckatkins commented:
Also fwiw the unit tests pass using this with `platform-python` on rhel8

JBlaschke commented on share/spack/setup-env.fish:
```diff
@@ -667,6 +667,19 @@ set -l sp_source_file (status -f)  # name of current file
+# Identify and lock the python interpreter
+#
+for cmd in "$SPACK_PYTHON" python3 python python2
+    set -l _sp_python (command -s "$cmd")
```
For consistency with POSIX shells, we should consider using `command -v` instead of `-s` here:
> For basic compatibility with POSIX command, the -v flag is recognized as an alias for -s
(fish `command` man-pages)

tgamblin replied:
I think we can do this one later since it's just in the `fish` script.
